interfaces/system_key.go

```diff
@@ -20,47 +20,86 @@
 package interfaces
 
 import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
 	"os"
 	"path/filepath"
-
-	"gopkg.in/yaml.v2"
+	"reflect"
+	"strings"
 
 	"github.com/snapcore/snapd/dirs"
 	"github.com/snapcore/snapd/logger"
 	"github.com/snapcore/snapd/osutil"
 	"github.com/snapcore/snapd/release"
 )
 
+// ErrSystemKeyIncomparableVersions indicates that the system-key
+// on disk and the system-key calculated from generateSystemKey
+// have different inputs and are therefor incomparable.
+//
+// This means:
+// - "snapd" needs to re-generate security profiles
+// - "snap run" cannot wait for those security profiles
+var (
+	ErrSystemKeyIncomparableVersions = errors.New("system-key versions not comparable")
+	ErrSystemKeyMissing              = errors.New("system-key missing on disk")
+)
+
 // systemKey describes the environment for which security profiles
 // have been generated. It is useful to compare if the current
 // running system is similar enough to the generated profiles or
 // if the profiles need to be re-generated to match the new system.
+//
+// Note that this key gets generated on *each* `snap run` - so it
+// *must* be cheap to calculate it (no hashes of big binaries etc).
 type systemKey struct {
-	BuildID          string   `yaml:"build-id"`
-	AppArmorFeatures []string `yaml:"apparmor-features"`
-	NFSHome          bool     `yaml:"nfs-home"`
-	OverlayRoot      string   `yaml:"overlay-root"`
-	Core             string   `yaml:"core,omitempty"`
-	SecCompActions   []string `yaml:"seccomp-features"`
+	// IMPORTANT: when adding new inputs bump this version
+	Version int `json:"version"`
+
+	BuildID          string   `json:"build-id"`
```

Comment:
Can you please add a newline below BuildID and add this comment
```
// The elements below are external to snapd and change depending on
// the running kernel and system configuration.
```

Reply:
Having a comment here is good, but that comment is saying the obvious: that these settings change when the system changes.

I think what we need to say is that these settings are here because they may cause profiles to be built in a different way, which may affect the ability of snap applications to run correctly after snapd (and thus snap run/snap-confine/etc) are updated, but before the profiles are generated.

Reply:
I updated the comment now, hope it adds clarity.